File: src/lib/stats.ts
export interface DataPoint {
  date: string;
  value: number;
}

export function mean(values: number[]): number {
  return values.reduce((a, b) => a + b, 0) / values.length;
}

export function stddev(values: number[]): number {
  const m = mean(values);
  const variance = values.reduce((acc, v) => acc + (v - m) ** 2, 0) / values.length;
  return Math.sqrt(variance);
}

export function zscore(value: number, values: number[]): number {
  const m = mean(values);
  const s = stddev(values);
  if (s === 0) return 0;
  return (value - m) / s;
}

// Rolling correlation between two aligned time series (same dates assumed)
export function rollingCorrelation(
  a: number[],
  b: number[],
  window: number
): (number | null)[] {
  const result: (number | null)[] = [];
  for (let i = 0; i < a.length; i++) {
    if (i < window - 1) {
      result.push(null);
      continue;
    }
    const sliceA = a.slice(i - window + 1, i + 1);
    const sliceB = b.slice(i - window + 1, i + 1);
    result.push(pearson(sliceA, sliceB));
  }
  return result;
}

function pearson(a: number[], b: number[]): number {
  const n = a.length;
  const ma = mean(a);
  const mb = mean(b);
  let num = 0,
    da = 0,
    db = 0;
  for (let i = 0; i < n; i++) {
    num += (a[i] - ma) * (b[i] - mb);
    da += (a[i] - ma) ** 2;
    db += (b[i] - mb) ** 2;
  }
  const denom = Math.sqrt(da * db);
  return denom === 0 ? 0 : num / denom;
}

// Align two series by date — returns matched pairs
export function alignByDate(
  seriesA: DataPoint[],
  seriesB: DataPoint[]
): { date: string; a: number; b: number }[] {
  const mapB = new Map(seriesB.map((p) => [p.date, p.value]));
  return seriesA
    .filter((p) => mapB.has(p.date))
    .map((p) => ({ date: p.date, a: p.value, b: mapB.get(p.date)! }));
}

// Align three series by date
export function alignThree(
  a: DataPoint[],
  b: DataPoint[],
  c: DataPoint[]
): { date: string; a: number; b: number; c: number }[] {
  const mapB = new Map(b.map((p) => [p.date, p.value]));
  const mapC = new Map(c.map((p) => [p.date, p.value]));
  return a
    .filter((p) => mapB.has(p.date) && mapC.has(p.date))
    .map((p) => ({ date: p.date, a: p.value, b: mapB.get(p.date)!, c: mapC.get(p.date)! }));
}

File: backend/lib/stats.py
"""Statistical utilities — mirrors stats.ts, implemented with numpy."""
from typing import Optional

import numpy as np


def mean(values: list[float]) -> float:
    return float(np.mean(values))


def stddev(values: list[float]) -> float:
    return float(np.std(values))


def zscore(value: float, values: list[float]) -> float:
    arr = np.array(values, dtype=float)
    s = arr.std()
    if s == 0:
        return 0.0
    return float((value - arr.mean()) / s)


def rolling_correlation(
    a: list[float], b: list[float], window: int
) -> list[Optional[float]]:
    """Pearson correlation over a sliding window. Returns None for positions < window."""
    result: list[Optional[float]] = []
    arr_a = np.array(a, dtype=float)
    arr_b = np.array(b, dtype=float)
    n = len(arr_a)
    for i in range(n):
        if i < window - 1:
            result.append(None)
            continue
        sa = arr_a[i - window + 1 : i + 1]
        sb = arr_b[i - window + 1 : i + 1]
        result.append(_pearson(sa, sb))
    return result


def _pearson(a: np.ndarray, b: np.ndarray) -> Optional[float]:
    ma, mb = a.mean(), b.mean()
    num = float(((a - ma) * (b - mb)).sum())
    denom = float(np.sqrt(((a - ma) ** 2).sum() * ((b - mb) ** 2).sum()))
    if denom == 0:
        return 0.0
    return num / denom


def align_by_date(
    series_a: list[dict], series_b: list[dict]
) -> list[dict]:
    """Inner join two [{date, value}] series on date."""
    map_b = {p["date"]: p["value"] for p in series_b}
    return [
        {"date": p["date"], "a": p["value"], "b": map_b[p["date"]]}
        for p in series_a
        if p["date"] in map_b
    ]


def align_three(
    a: list[dict], b: list[dict], c: list[dict]
) -> list[dict]:
    """Inner join three [{date, value}] series on date."""
    map_b = {p["date"]: p["value"] for p in b}
    map_c = {p["date"]: p["value"] for p in c}
    return [
        {"date": p["date"], "a": p["value"], "b": map_b[p["date"]], "c": map_c[p["date"]]}
        for p in a
        if p["date"] in map_b and p["date"] in map_c
    ]
